fleetspeak/src/server/stats.go

```diff
@@ -142,50 +221,66 @@ var (
 type PrometheusStatsCollector struct{}
 
 func (s PrometheusStatsCollector) MessageIngested(backlogged bool, m *fspb.Message) {
-	messagesIngested.Inc()
-	if backlogged {
-		backlogMessagesFromDatastore.Inc()
-	} else {
-		messagesReceivedFromClient.Inc()
-	}
+	messagesIngested.WithLabelValues(strconv.FormatBool(backlogged), m.Source.ServiceName, m.Destination.ServiceName, m.MessageType).Inc()
 }
 
 func (s PrometheusStatsCollector) MessageSaved(service, messageType string, forClient bool, savedPayloadBytes int) {
-	messagesSaved.Inc()
-	messagesSavedSize.Add(float64(savedPayloadBytes))
+	messagesSaved.WithLabelValues(service, messageType, strconv.FormatBool(forClient)).Inc()
+	messagesSavedSize.WithLabelValues(service, messageType, strconv.FormatBool(forClient)).Add(float64(savedPayloadBytes))
 }
 
 func (s PrometheusStatsCollector) MessageProcessed(start, end time.Time, service, messageType string) {
-	messagesProcessed.Inc()
+	messagesProcessed.WithLabelValues(messageType).Observe(end.Sub(start).Seconds())
 }
 
 func (s PrometheusStatsCollector) MessageErrored(start, end time.Time, service, messageType string, isTemp bool) {
-	messagesErrored.Inc()
-	if isTemp {
-		messagesErroredTemp.Inc()
-	} else {
-		messagesErroredPerm.Inc()
-	}
+	messagesErrored.WithLabelValues(messageType, strconv.FormatBool(isTemp)).Observe(end.Sub(start).Seconds())
 }
 
 func (s PrometheusStatsCollector) MessageDropped(service, messageType string) {
-	messagesDropped.Inc()
+	messagesDropped.WithLabelValues(service, messageType).Inc()
 }
 
 func (s PrometheusStatsCollector) ClientPoll(info stats.PollInfo) {
-	clientPolls.Inc()
+	clientPolls.WithLabelValues(strconv.Itoa(info.Status), info.Type.String(), strconv.FormatBool(info.CacheHit)).Inc()
```

Comment:
There's a lot of repetition here. I'd put label values in variables and reuse them without calling all the functions again?